src/acl/external/eDirectory_userip/ext_edirectory_userip_acl.cc

```diff
@@ -1098,48 +995,42 @@ SearchFilterLDAP(edui_ldap_t *l, char *group)
-            int ln = snprintf(bufd, sizeof(bufd), "(networkAddress=10\\23\\00\\00%s)(networkAddress=11\\23\\00\\00%s))", \
-                              bufc, bufc);
-            strncat(bufb, bufd, ln);
-        } else
-            strncat(bufb, ")", 1);
-        strncat(bufa, bufb, strlen(bufb));
-        strncat(bufa, ")", 1);
+            int ln = snprintf(bufd, sizeof(bufd), "(networkAddress=10\\23\\00\\00%s)(networkAddress=11\\23\\00\\00%s)", bufc, bufc);
+            if (ln < 0 || sizeof(bufd) <= static_cast<size_t>(ln))
+                return LDAP_ERR_OOB;
+        }
+        int x = snprintf(bufa, sizeof(bufa), "(&%s(|(networkAddress=1\\23%s)%s))", edui_conf.search_filter, bufc, bufd);
+        if (x < 0 || sizeof(bufa) <= static_cast<size_t>(x))
+            return LDAP_ERR_OOB;
+
     } else {
         /* Needs groupMembership= to add... */
-        xstrncpy(bufa, "(&(&", sizeof(bufa));
-        strncat(bufa, edui_conf.search_filter, strlen(edui_conf.search_filter));
         /* groupMembership -- NOTE: Squid *MUST* provide "cn=" from squid.conf */
-        snprintf(bufg, sizeof(bufg), "(groupMembership=%s", group);
         if ((l->basedn[0] != '\0') && (strstr(group, l->basedn) == NULL)) {
-            strncat(bufg, ",", 1);
-            strncat(bufg, l->basedn, strlen(l->basedn));
+            int ln = snprintf(bufg, sizeof(bufg), ",%s", l->basedn);
+            if (ln < 0 || sizeof(bufd) <= static_cast<size_t>(ln))
+                return LDAP_ERR_OOB;
         }
-        strncat(bufg, ")", 1);
-        strncat(bufa, bufg, strlen(bufg));
         /* networkAddress */
-        snprintf(bufb, sizeof(bufb), "(|(networkAddress=1\\23%s)", bufc);
         if (l->status & LDAP_IPV4_S) {
-            int ln = snprintf(bufd, sizeof(bufd), "(networkAddress=8\\23\\00\\00%s)(networkAddress=9\\23\\00\\00%s))", \
-                              bufc, bufc);
-            strncat(bufb, bufd, ln);
+            int ln = snprintf(bufd, sizeof(bufd), "(networkAddress=8\\23\\00\\00%s)(networkAddress=9\\23\\00\\00%s)", bufc, bufc);
+            if (ln < 0 || sizeof(bufd) <= static_cast<size_t>(ln))
+                return LDAP_ERR_OOB;
         } else if (l->status & LDAP_IPV6_S) {
-            int ln = snprintf(bufd, sizeof(bufd), "(networkAddress=10\\23\\00\\00%s)(networkAddress=11\\23\\00\\00%s))", \
-                              bufc, bufc);
-            strncat(bufb, bufd, ln);
-        } else
-            strncat(bufb, ")", 1);
-        strncat(bufa, bufb, strlen(bufb));
-        strncat(bufa, "))", 2);
+            int ln = snprintf(bufd, sizeof(bufd), "(networkAddress=10\\23\\00\\00%s)(networkAddress=11\\23\\00\\00%s)", bufc, bufc);
+            if (ln < 0 || sizeof(bufd) <= static_cast<size_t>(ln))
+                return LDAP_ERR_OOB;
+        }
+        int x = snprintf(bufa, sizeof(bufa), "(&(&%s(groupMembership=%s%s)(|(networkAddress=1\\23%s)%s)))", edui_conf.search_filter, group, bufg, bufc, bufd);
```

Comment:
const, auto

Reply:
Do you expect snprintf() return type to be changing in future? if not using the correct type is more beneficial - it is not exactly complex or longer than 4 chars to type. More importantly auto hides the info about whether static_cast is safe or not (eg not a pointer type being cast to size_t value).

Reply:
> Do you expect snprintf() return type to be changing in future?

Wrong question: I do not (want to) know what the return type is (or will be)! As you know, it is easy to type the wrong type in similar contexts. Using `auto` avoids those problems.

> auto hides the info about whether static_cast is safe or not (eg not a pointer type being cast to size_t value).

[static_cast](http://en.cppreference.com/w/cpp/language/static_cast) does not do the "unsafe" casts you are probably thinking about. In this context, static_cast is similar to an implicit conversion. Besides, I believe it is possible to just make the result unsigned (without casting/changing the underlying integer type), and it is probably a good idea to add a wrapper for doing so in a readable way. As you know, it is easy to screw up manual post-call checks in such contexts.